File: src/main/java/net/lifecity/mc/skillmaster/skill/separatedskills/leafflow/LeafFlow.java
package net.lifecity.mc.skillmaster.skill.separatedskills.leafflow;

import net.lifecity.mc.skillmaster.SkillMaster;
import net.lifecity.mc.skillmaster.skill.SeparatedSkill;
import net.lifecity.mc.skillmaster.skill.SkillType;
import net.lifecity.mc.skillmaster.user.SkillUser;
import net.lifecity.mc.skillmaster.weapon.Weapon;
import org.bukkit.Material;
import org.bukkit.Particle;
import org.bukkit.Sound;
import org.bukkit.scheduler.BukkitRunnable;
import org.bukkit.util.Vector;

/**
 * 前方に突進しながら敵を攻撃するスキル
 */
import java.util.Arrays;

public class LeafFlow extends SeparatedSkill {

    private final double movePower;
    private final double yPower;
    private final double attackRadius;
    private final double damage;
    private final double attackImpact;
    private final double yImpact;

    /**
     * リーフフローのコンストラクタ
     * @param weapon 使用する武器
     * @param num スキルの番号
     * @param point 消費ポイント
     * @param activationTime 発動時間
     * @param interval インターバル
     * @param user 使用するユーザー
     * @param movePower 突進の強さ
     * @param yPower 突進のY軸への強さ
     * @param attackRadius 攻撃範囲
     * @param damage ダメージ
     * @param attackImpact ノックバック
     * @param yImpact Y軸へのノックバック
     */
    protected LeafFlow(
            Weapon weapon,
            int num,
            int point,
            int activationTime,
            int interval,
            SkillUser user,
            double movePower,
            double yPower,
            double attackRadius,
            double damage,
            double attackImpact,
            double yImpact
    ) {
        super(
                "リーフフロー",
                weapon,
                SkillType.ATTACK,
                Arrays.asList(
                        "前に進みながら斬撃します。",
                        "1回目の入力で前に移動します。",
                        "2回目の入力で攻撃します。"
                ),
                num,
                point,
                activationTime,
                interval,
                user
        );
        this.movePower = movePower;//1
        this.yPower = yPower;//0.15
        this.attackRadius = attackRadius;//1.8
        this.damage = damage;//3
        this.attackImpact = attackImpact;//1
        this.yImpact = yImpact;//0.15
    }

    @Override
    public void activate() {
        super.activate();

        Vector vector = user.getPlayer().getEyeLocation().getDirection()
                .normalize()
                .multiply(movePower)
                .setY(yPower);

        user.getPlayer().setVelocity(vector);

        // [LE]葉の流れ
        new BukkitRunnable() {
            int count = 0;

            @Override
            public void run() {
                if (!activated)
                    cancel();
                if (count >= 10)
                    cancel();
                if (user.getPlayer().getVelocity().length() <= 0.3)
                    cancel();

                particle(Particle.FALLING_DUST, user.getPlayer().getEyeLocation(), Material.SPRUCE_LEAVES.createBlockData());

                count++;
            }
        }.runTaskTimer(SkillMaster.instance, 0, 1);
    }

    @Override
    public void additionalInput() {

        // 一番近いEntityを攻撃
        boolean b = user.attackNearest(
                attackRadius,
                damage,
                user.getPlayer().getVelocity().normalize().multiply(attackImpact).setY(yImpact),
                Sound.ENTITY_PLAYER_ATTACK_SWEEP
        );

        // LE
        if (b) {
            particle(Particle.EXPLOSION_LARGE, user.getNearEntities(attackRadius).get(0).getLocation().add(0, 2, 0));
            for (int i = 0; i < 6; i++) {
                particle(Particle.FLAME, user.getPlayer().getLocation());
            }
        }

        // 終了
        deactivate();
    }
}

File: src/main/java/net/lifecity/mc/skillmaster/skill/compositeskills/LeafFlow.kt
package net.lifecity.mc.skillmaster.skill.compositeskills

import com.github.syari.spigot.api.particle.spawnParticle
import com.github.syari.spigot.api.scheduler.runTaskTimer
import com.github.syari.spigot.api.sound.playSound
import net.lifecity.mc.skillmaster.SkillMaster
import net.lifecity.mc.skillmaster.skill.*
import net.lifecity.mc.skillmaster.user.SkillUser
import net.lifecity.mc.skillmaster.utils.NearTargets
import net.lifecity.mc.skillmaster.weapon.Weapon
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.Particle
import org.bukkit.Sound
import java.util.*

/**
 * 前方に突進しながら敵を攻撃するスキル
 */
class LeafFlow(
    override val activationTime: Int = 8,
    override val canCancel: Boolean = true,
    override val name: String = "リーフフロー",
    override val weaponList: List<Weapon> = listOf(Weapon.STRAIGHT_SWORD, Weapon.DAGGER, Weapon.RAPIER),
    override val type: SkillType = SkillType.ATTACK,
    override val lore: List<String> = listOf(
        "前に進みながら斬撃します。",
        "1回目の入力で前に移動します。",
        "2回目の入力で攻撃します。"
    ),
    override var isActivated: Boolean = false,
    override val interval: Int = 100,
    override var inInterval: Boolean = false,
    override val user: SkillUser
) : AttackSkill(), ICompositeSkill {

    val skillController = SkillController(this)

    override fun onAdditionalInput() {
        // 一番近いEntityを攻撃
        val target = NearTargets.search(user.player, 1.8) ?: return

        attackAddVector(user, target, 4.0,  user.player.velocity.normalize().multiply(1).setY(0.15))

        repeat(2) {
            target.location.add(0.0, 2.0, 0.0).spawnParticle(Particle.SWEEP_ATTACK)
        }
        repeat(5) {
            val randomLocation = randomLocation(0.3)
            randomLocation.add(user.player.location).spawnParticle(Particle.FLAME)
        }
        // 終了
        skillController.deactivate()
    }

    override fun register() {
        SkillManager(this).register()
    }

    override fun canActivate(): Boolean = true

    override fun onActivate() {
        val vector = user.player.eyeLocation.direction
            .normalize()
            .multiply(1.3)
            .setY(0.15)
        user.player.velocity = vector

        // ランダムロケーションを生成
        val leaves: MutableList<Location> = ArrayList()
        repeat(11) {
            randomLocation(0.25).let { leaves.add(it) }
        }
        val wind: MutableList<Location> = ArrayList()
        repeat (3) {
            randomLocation(0.5).let { wind.add(it) }
        }

        // エフェクト葉の流れ
        var count = 0
        SkillMaster.INSTANCE.runTaskTimer(1) {
            if (!isActivated) cancel()
            if (count >= 10) cancel()
            if (user.player.velocity.length() <= 0.3) cancel()

            for (location in leaves) {
                val particleLocation = user.player.location
                val loc = particleLocation.add(0.0, 1.0, 0.0).add(location)

                loc.spawnParticle(Particle.FALLING_DUST, data = Material.SPRUCE_LEAVES.createBlockData())
            }

            for (location in wind) {
                val particleLocation = user.player.location
                val loc = particleLocation.add(0.0, 1.0, 0.0).add(location)

                loc.spawnParticle(Particle.ELECTRIC_SPARK)
            }

            if (count % 2 == 0)
                user.player.location.playSound(Sound.BLOCK_WET_GRASS_BREAK)
            count++
        }
    }

    override fun onDeactivate() {
    }

    private fun randomLocation(max: Double): Location {
        val random = Random()
        var x = random.nextDouble() * max
        if (random.nextBoolean()) x *= -1.0
        var y = random.nextDouble() * max
        if (random.nextBoolean()) y *= -1.0
        var z = random.nextDouble() * max
        if (random.nextBoolean()) z *= -1.0
        return Location(user.player.world, x, y, z)
    }

}